redsky/streams.py

```diff
@@ -0,0 +1,630 @@
+        Stream.__init__(self, children=children)
+
+    def update(self, x, who=None):
+        if self.missing and who in self.missing:
+            self.missing.remove(who)
+
+        self.last[self.children.index(who)] = x
+        if not self.missing:
+            tup = tuple(self.last)
+            if tup and hasattr(tup[0], '__stream_merge__'):
+                tup = tup[0].__stream_merge__(*tup[1:])
+            return self.emit(tup)
+
+
+class concat(Stream):
+    def update(self, x, who=None):
+        L = []
+        for item in x:
+            y = self.emit(item)
+            if type(y) is list:
+                L.extend(y)
+            else:
+                L.append(y)
+        return L
+
+
+class unique(Stream):
+    def __init__(self, child, history=None, key=identity):
+        self.seen = dict()
+        self.key = key
+        if history:
+            from zict import LRU
+            self.seen = LRU(history, self.seen)
+
+        Stream.__init__(self, child)
+
+    def update(self, x, who=None):
+        y = self.key(x)
+        if y not in self.seen:
+            self.seen[y] = 1
+            return self.emit(x)
+
+
+class union(Stream):
+    def update(self, x, who=None):
+        return self.emit(x)
+
+
+class collect(Stream):
+    def __init__(self, child, cache=None):
+        if cache is None:
+            cache = deque()
+        self.cache = cache
+
+        Stream.__init__(self, child)
+
+    def update(self, x, who=None):
+        self.cache.append(x)
+
+    def flush(self, _=None):
+        out = tuple(self.cache)
+        self.emit(out)
+        self.cache.clear()
+
+
+class starmap(Stream):
+    def __init__(self, func, child, raw=False, **kwargs):
+        self.func = func
+        self.kwargs = kwargs
+        self.raw = raw
+
+        Stream.__init__(self, child)
+
+    def update(self, x, who=None):
+        if not self.raw and hasattr(x, '__stream_map__'):
+            result = x.__stream_map__(self.func, **self.kwargs)
+        else:
+            result = self.func(*x, **self.kwargs)
+
+        return self.emit(result)
+
+
+class dstarmap(Stream):
+    def __init__(self, func, child, raw=False, output_info=None,
+                 input_info=None, **kwargs):
+        self.func = func
+        self.kwargs = kwargs
+        self.raw = raw
+
+        Stream.__init__(self, child, output_info=output_info,
+                        input_info=input_info, **kwargs)
+        ### TODO: update the provenence now that we have the func
+
+    def update(self, x, who=None):
+        # massage the pair(s)
+        res = self.dispatch(x)
+        # if we are giving back a new doc, just emit it
+        if isinstance(res, tuple) and res[0] in ['start', 'descriptor',
+                                                 'stop']:
+            return self.emit(res)
+        try:
+            # we need to expose the raw event data
+            res = self.event_guts(res)
+            if not self.raw and hasattr(x, '__stream_map__'):
```

Comment:
hmm, seems this could be a problem actually.

@mrocklin was a step ahead, and this could be solved through the use of single dispatch [here](https://github.com/mrocklin/streams/pull/13/files#diff-3b74a983c05dda1f6f13bc3301cb0288R10).

However, the dispatch would have to be run on a tuple, since that's what the current event model returns (correct me if wrong):
```python
@stream_map.register(tuple)
def stream_map(elem, func **kwargs):
    if elem[0] == 'start':
    ...
    else:
        # if none of the above return as is
        return func(elem, **kwargs)
```
Since `stream_map` is actually run recursively on further returned data elements, if any subelement happens to return a tuple that satisfies condition in the dispatch, then if result in some unwanted results.
Probably for this to work we'd need to wrap the incoming `tuple` as a new unique class which we can dispatch on, or include some sentinel in the tuple that we check for (i.e. check `isinstance(elem[1], Document)` or `hasattr(elem[0], 'start')` etc).

@danallan, is there some standard way to check and guarantee that an incoming tuple is a name,doc pair?